### PR TITLE
oop-special_class_method-hard-q10

### PR DESCRIPTION
**Pull request by KaiKhademi**
## PR Checklist
- [x]  Are helper methods really needed here?
    > Opt for hardcoding input/output into the input provider when possible
- [x]  Are the assertion fail messages good?
    - [x]  Logical (does it make sense?)
    - [x]  Readable (can a student understand what’s wrong?)
    - [x]  Not too feed-y (does it give away the answer?)
    - [x]  Grammatically sound (is the grammar and punctuation in the sentence proper?)
    - [x]  Specific (does the fail message relate to the question?)
- [x]  Is the code formatted well?
    > Run `CTRL` + `ALT` + `L` in IntelliJ
- [x]  EOF newlines
- [x]  Do the comments for Parsons follow the correct convention?
- [x]  Is the question being tested sufficiently?
    - [x]  Are there representative cases to test the program?
    - [x]  What about edge cases? (empty strings, negatives, infinity, off-by-one errors, etc)
        > If the question doesn’t specify what to do, mark it as a bad question
- [x]  Are the test method names good?
    > Test method names show up in a “What went well” section on the website, so “mathTest” will become “What went well: math test”, which doesn’t make much sense. A better name in this example would be “calculatesTaxCorrectly” so it shows up as “What went well: calculates tax correctly”

### CODE REVIEW

File: src/oop/special_class_method/hard/q10/QuestionText.md [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+    - a `getAwardRecipient` method which is accessible from other classes and returns the value of the **
+      awardRecipient** attribute
+    - a `setAwardCategory` method which accepts a `String` value, updates the **awardCategory** attribute with it, and
+      is
+      accessible
+      from other classes
+    - a `getAwardCategory` method which is accessible from other classes and returns the value of the **awardCategory**
+      attribute
+    - a `toString` method which returns a String of the information about the `Award` class of the form
+      `"The award for " + awardCategory + " category goes to " + awardRecipient`.
```
Comment (SamamthaH) [inline src/oop/special_class_method/hard/q10/QuestionText.md:36]: for consistency this should be x y format

File: src/oop/special_class_method/hard/q10/QuestionText.md [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+
+- setAwardRecipient(awardRecipient);
+- setAwardCategory(awardCategory);
+- return AwardCategory;
+- return awardCategory = awardCategory;
+- public String awardCategory;
+- public String awardRecipient;
+- private class Award
+- public Award()
+- public String toString()
```
Comment (SamamthaH) [inline src/oop/special_class_method/hard/q10/QuestionText.md:48]: This will not work as a distractor if the person entering the question on the website puts the opening and closing brackets on their own lines (as I do)

File: src/oop/special_class_method/hard/q10/Award.java [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+        this.awardRecipient = awardRecipient;
+        this.awardCategory = awardCategory;
+    }
+
+    public void setAwardRecipient(String awardRecipient) {
+        this.awardRecipient = awardRecipient;
+    }
+
+    public String getAwardRecipient() {
+        return this.awardRecipient;
```
Comment (SamamthaH) [inline src/oop/special_class_method/hard/q10/Award.java:17]: the "this" is not needed for the get methods

File: src/oop/special_class_method/hard/q10/MainTest.java [hunk 4]
```diff
@@ -0,0 +1,192 @@
...
+    @ParameterizedTest
+    @MethodSource("inputProvider")
+    public void awardClassHasCorrectConstructor(String value1, String value2) throws Throwable {
+        Object[][] arguments = {
+                {value1, String.class},
+                {value2, String.class}
+        };
+        Class<?>[] constructorClasses = {String.class, String.class};
+        Object classInstance = testClass.createInstance(arguments);
+        String incorrectDoubleValueMessage =
```
Comment (SamamthaH) [inline src/oop/special_class_method/hard/q10/MainTest.java:63]: this should be labelled String not double

File: src/oop/special_class_method/hard/q10/MainTest.java [hunk 5]
```diff
@@ -0,0 +1,192 @@
...
+                {initialValue, String.class},
+                {"", String.class}
+        };
+        Object classInstance = testClass.createInstance(arguments);
+        Object[][] setSizeArguments = {
+                {setValue, String.class}
+        };
+        String[] setMethodModifiers = {"public"};
+        String incorrectSetMethodMessage = String.join(" ",
+                "Your", setFirstMethodName, "does not correctly set the value of the", firstFieldName, "attribute.");
```
Comment (SamamthaH) [inline src/oop/special_class_method/hard/q10/MainTest.java:89]: make sure all the error messages where you include a method name has "method" after it

File: src/oop/special_class_method/hard/q10/MainTest.java [hunk 6]
```diff
@@ -0,0 +1,192 @@
...
+
+    @ParameterizedTest
+    @MethodSource("inputProvider")
+    public void awardClassHasCorrectSetAwardCategoryMethod(String initialValue, String updatedValue) throws Throwable {
+        Object[][] arguments = {
+                {"", String.class},
+                {initialValue, String.class}
+        };
+        Object classInstance = testClass.createInstance(arguments);
+        testClass.setFieldValue(classInstance, initialValue, secondFieldName);
```
Comment (SamamthaH) [inline src/oop/special_class_method/hard/q10/MainTest.java:118]: why is this here?

File: src/oop/special_class_method/hard/q10/MainTest.java [hunk 7]
```diff
@@ -0,0 +1,192 @@
...
+
+    @ParameterizedTest
+    @MethodSource("inputProvider")
+    public void awardClassHasCorrectGetAwardCategoryMethod(String value1, String value2) throws Throwable {
+        Object[][] arguments = {
+                {value1, String.class},
+                {value2, String.class}
+        };
+        Object classInstance = testClass.createInstance(arguments);
+        testClass.setFieldValue(classInstance, value2, secondFieldName);
```
Comment (SamamthaH) [inline src/oop/special_class_method/hard/q10/MainTest.java:138]: why is this here?

File: src/oop/special_class_method/hard/q10/MainTest.java [hunk 8]
```diff
@@ -0,0 +1,192 @@
...
+        _assertEquals("initialValue", testClass.callMethod(getFirstMethodName, classInstance),
+                "Your " + getFirstMethodName + " method does not return the correct value.");
+        _assertEquals("secondValue", testClass.callMethod(getSecondMethodName, classInstance),
+                "Your " + getSecondMethodName + " method does not return the correct value.");
+        Object[][] setDoubleArguments = {
+                {value, String.class}
+        };
+        testClass.callMethod(setFirstMethodName, setDoubleArguments, classInstance);
+        _assertEquals(value, testClass.callMethod(getFirstMethodName, classInstance),
+                "Your " + getFirstMethodName + " method does not return the correct value calling the " + setFirstMethodName + " method.");
```
Comment (SamamthaH) [inline src/oop/special_class_method/hard/q10/MainTest.java:179]:
```suggestion
                "Your " + getFirstMethodName + " method does not return the correct value after calling the " + setFirstMethodName + " method.");
```

File: src/oop/special_class_method/hard/q10/MainTest.java [hunk 9]
```diff
@@ -0,0 +1,192 @@
...
+        Object[][] setBooleanArguments = {
+                {b, String.class}
+        };
+        testClass.callMethod(setSecondMethodName, setBooleanArguments, classInstance);
+        _assertEquals(b, testClass.callMethod(getSecondMethodName, classInstance),
+                "Your " + getSecondMethodName + " method does not return the correct value calling the " + setSecondMethodName + " method.");
```
Comment (SamamthaH) [inline src/oop/special_class_method/hard/q10/MainTest.java:185]:
```suggestion
                "Your " + getSecondMethodName + " method does not return the correct value after calling the " + setSecondMethodName + " method.");
```

**Review by SamamthaH — CHANGES_REQUESTED** [review]

File: src/oop/special_class_method/hard/q10/QuestionText.md [hunk 10]
```diff
@@ -0,0 +1,45 @@
...
+    - a `setAwardRecipient` method which accepts a `String` value, updates the **awardRecipient** attribute with it, and
+      is accessible from other classes
+    - a `getAwardRecipient` method which is accessible from other classes and returns the value of the **
+      awardRecipient** attribute
+    - a `setAwardCategory` method which accepts a `String` value, updates the **awardCategory** attribute with it, and
+      is accessible from other classes
+    - a `getAwardCategory` method which is accessible from other classes and returns the value of the **awardCategory**
+      attribute
+    - a `toString` method which returns a String of the information about the `Award` class of the form
+      `"The award for " + x + " category goes to " + y`. where is is the value of **awardCategory** attribute and y
```
Comment (SamamthaH) [inline src/oop/special_class_method/hard/q10/QuestionText.md:32]:
```suggestion
      `"The award for x category goes to y"`. where x is the value of **awardCategory** attribute and y
```

File: src/oop/special_class_method/hard/q10/MainTest.java [hunk 11]
```diff
@@ -0,0 +1,190 @@
...
+    public void awardClassHasCorrectToStringMethod(String value, String b) throws Throwable {
+        Object[][] arguments = {
+                {value, String.class},
+                {b, String.class}
+        };
+        Object classInstance = testClass.createInstance(arguments);
+        String[] methodModifiers = {"public"};
+        String expected = "The award for " + b + " category goes to " + value;
+        String incorrectToStringMessage = String.join(" ",
+                "Your", className, " toString method does not return the correct String.");
```
Comment (SamamthaH) [inline src/oop/special_class_method/hard/q10/MainTest.java:155]:
```suggestion
                "Your", className, "toString method does not return the correct String.");
```

File: src/oop/special_class_method/hard/q10/QuestionText.md [hunk 12]
```diff
@@ -0,0 +1,45 @@
...
+    - A `String` attribute named **awardCategory** which is not accessible from other classes
+
+- Constructors:
+    - A constructor which is accessible from other classes and accepts two `String` parameter to initialize the
+      **awardRecipient** and **awardCategory** attributes in that order.
+
+- Methods:
+    - a `setAwardRecipient` method which accepts a `String` value, updates the **awardRecipient** attribute with it, and
+      is accessible from other classes
+    - a `getAwardRecipient` method which is accessible from other classes and returns the value of the **
```
Comment (SamamthaH) [inline src/oop/special_class_method/hard/q10/QuestionText.md:25]: the ** formatting doesn't transfer between lines

File: src/oop/special_class_method/hard/q10/QuestionText.md [hunk 13]
```diff
@@ -0,0 +1,45 @@
...
+    - a `getAwardCategory` method which is accessible from other classes and returns the value of the **awardCategory**
+      attribute
+    - a `toString` method which returns a String of the information about the `Award` class of the form
+      `"The award for " + x + " category goes to " + y`. where is is the value of **awardCategory** attribute and y
+      is the value of **awardRecipient** attribute
+
+### Distractors:
+
+- setAwardRecipient(awardRecipient);
+- setAwardCategory(awardCategory);
```
Comment (SamamthaH) [inline src/oop/special_class_method/hard/q10/QuestionText.md:38]: these two distractors could potentially be used in the constructor

File: src/oop/special_class_method/hard/q10/MainTest.java [hunk 14]
```diff
@@ -0,0 +1,190 @@
...
+        Object classInstance = testClass.createInstance(arguments);
+        Object[][] setSizeArguments = {
+                {setValue, String.class}
+        };
+        String[] setMethodModifiers = {"public"};
+        String incorrectSetMethodMessage = String.join(" ",
+                "Your", setFirstMethodName, "method does not correctly set the value of the", firstFieldName, "attribute.");
+        Object setMethodOutput = testClass.callMethod(setFirstMethodName, setSizeArguments, setMethodModifiers, classInstance);
+        _assertEquals(setValue, testClass.getFieldValue(classInstance, firstFieldName), incorrectSetMethodMessage);
+        assertNull(setMethodOutput, String.join(" ", "Your", setFirstMethodName, "should not return any output"));
```
Comment (SamamthaH) [inline src/oop/special_class_method/hard/q10/MainTest.java:92]:
```suggestion
        assertNull(setMethodOutput, String.join(" ", "Your", setFirstMethodName, "method should not return any output"));
```

File: src/oop/special_class_method/hard/q10/MainTest.java [hunk 15]
```diff
@@ -0,0 +1,190 @@
...
+        Object classInstance = testClass.createInstance(arguments);
+        Object[][] setMethodArguments = {
+                {updatedValue, String.class}
+        };
+        String[] setMethodModifiers = {"public"};
+        String incorrectSetMethodMessage = String.join(" ",
+                "Your", setSecondMethodName, "method does not correctly set the value of the", secondFieldName, "attribute.");
+        Object setMethodOutput = testClass.callMethod(setSecondMethodName, setMethodArguments, setMethodModifiers, classInstance);
+        _assertEquals(updatedValue, testClass.getFieldValue(classInstance, secondFieldName), incorrectSetMethodMessage);
+        assertNull(setMethodOutput, String.join(" ", "Your", setSecondMethodName, "should not return any output"));
```
Comment (SamamthaH) [inline src/oop/special_class_method/hard/q10/MainTest.java:126]:
```suggestion
        assertNull(setMethodOutput, String.join(" ", "Your", setSecondMethodName, "method should not return any output"));
```

**Review by SamamthaH — CHANGES_REQUESTED** [review]

File: src/oop/special_class_method/hard/q10/QuestionText.md [hunk 16]
```diff
@@ -0,0 +1,42 @@
...
+    - a `setAwardCategory` method which accepts a `String` value, updates the **awardCategory** attribute with it, and
+      is accessible from other classes
+    - a `getAwardCategory` method which is accessible from other classes and returns the value of the **awardCategory**
+      attribute
+    - a `toString` method which returns a String of the information about the `Award` class of the form
+      `"The award for x category goes to y"`. where x is the value of **awardCategory** attribute and y
+      is the value of **awardRecipient** attribute
+
+### Distractors:
+- return AwardCategory;
```
Comment (SamamthaH) [inline src/oop/special_class_method/hard/q10/QuestionText.md:36]: since we've previously taken away distractors we should probably add more back in since there aren't a ton left

File: src/oop/special_class_method/hard/q10/MainTest.java [hunk 17]
```diff
@@ -0,0 +1,190 @@
...
+    private final String setFirstMethodName = "setAwardRecipient";
+    private ObjectTest testClass;
+
+    @BeforeEach
+    public void setUp() {
+        String classString = "oop.special_class_method.hard.q10." + className;
+        testClass = new ObjectTest(classString);
+    }
+
+    @Test
```
Comment (SamamthaH) [inline src/oop/special_class_method/hard/q10/MainTest.java:32]: modify to include the new hasField method

File: src/oop/special_class_method/hard/q10/MainTest.java [hunk 18]
```diff
@@ -0,0 +1,190 @@
...
+    public void awardClassHasCorrectAttributes() {
+        assertTrue(testClass.hasField(firstFieldName, String.class),
+                "Your " + className + " class is missing the " + firstFieldName + " attributes.");
```
Comment (SamamthaH) [inline src/oop/special_class_method/hard/q10/MainTest.java:35]: change "attributes" to singular here and on line 39

File: src/oop/special_class_method/hard/q10/MainTest.java [hunk 19]
```diff
@@ -0,0 +1,190 @@
...
+    private static Stream<Arguments> inputProvider() {
+        return Stream.of(
+                Arguments.of("First", "Second"),
+                Arguments.of("Science", "Anna"),
+                Arguments.of("Jack", "History"),
+                Arguments.of("Finn", "Fiction")
+
+        );
+    }
+
```
Comment (SamamthaH) [inline src/oop/special_class_method/hard/q10/MainTest.java:53]: There is no test checking if the constructor exists, only that it initializes the attributes correctly and then has the correct modifier

File: src/oop/special_class_method/hard/q10/MainTest.java [hunk 20]
```diff
@@ -0,0 +1,190 @@
...
+                "Your " + getFirstMethodName + " method does not return the correct value after calling the " + setFirstMethodName + " method.");
+        Object[][] setBooleanArguments = {
+                {b, String.class}
+        };
+        testClass.callMethod(setSecondMethodName, setBooleanArguments, classInstance);
+        _assertEquals(b, testClass.callMethod(getSecondMethodName, classInstance),
+                "Your " + getSecondMethodName + " method does not return the correct value after calling the " + setSecondMethodName + " method.");
+        String expectedToString = "The award for " + b + " category goes to " + value;
+        String incorrectToStringMessage = String.join(" ",
+                "Your", className, " toString method does not return the correct String after updating the values of its attributes using its setter methods.");
```
Comment (SamamthaH) [inline src/oop/special_class_method/hard/q10/MainTest.java:186]:
```suggestion
                "Your", className, "toString method does not return the correct String after updating the values of its attributes using its setter methods.");
```

**Review by SamamthaH — CHANGES_REQUESTED** [review]

File: src/oop/special_class_method/hard/q10/MainTest.java [hunk 21]
```diff
@@ -0,0 +1,199 @@
...
+                {"", String.class}
+        };
+        Object classInstance = testClass.createInstance(arguments);
+        Object[][] setSizeArguments = {
+                {setValue, String.class}
+        };
+        String[] setMethodModifiers = {"public"};
+        String incorrectSetMethodMessage = String.join(" ",
+                "Your", setFirstMethodName, "method does not correctly set the value of the", firstFieldName, "attribute.");
+        Object setMethodOutput = testClass.callMethod(setFirstMethodName, setSizeArguments, setMethodModifiers, classInstance);
```
Comment (SamamthaH) [inline src/oop/special_class_method/hard/q10/MainTest.java:99]: use .hasMethod first and void.type instead of assertNull here and everywhere else

File: src/oop/special_class_method/hard/q10/QuestionText.md [hunk 22]
```diff
@@ -0,0 +1,44 @@
+# oop-special_class_method-hard-q10
+
+## Question Text
+
+You've been chosen to make plans for an award show! Your first task is to model the awards being received, so create a
```
Comment (SamamthaH) [inline src/oop/special_class_method/hard/q10/QuestionText.md:5]: get rid of duplicate class information

File: src/oop/special_class_method/hard/q10/QuestionText.md [hunk 23]
```diff
@@ -0,0 +1,44 @@
...
+### Award Class:
+
+Your `Award` class must have the following items:
+
+- Attributes:
+    - A `String` attribute named **awardRecipient** which is not accessible from other classes
+    - A `String` attribute named **awardCategory** which is not accessible from other classes
+
+- Constructors:
+    - A constructor which is accessible from other classes and accepts two `String` parameter to initialize the
```
Comment (SamamthaH) [inline src/oop/special_class_method/hard/q10/QuestionText.md:19]:
```suggestion
    - A constructor which is accessible from other classes and accepts two `String` parameters to initialize the
```

File: src/oop/special_class_method/hard/q10/MainTest.java [hunk 24]
```diff
@@ -0,0 +1,199 @@
...
+    @BeforeEach
+    public void setUp() {
+        String classString = "oop.special_class_method.hard.q10." + className;
+        testClass = new ObjectTest(classString);
+    }
+
+    @Test
+    public void awardClassHasCorrectAttributes() {
+        assertTrue(testClass.hasField(firstFieldName),
+                "Your " + className + " class is missing the " + firstFieldName + " attributes, or it is spelled incorrectly.");
```
Comment (SamamthaH) [inline src/oop/special_class_method/hard/q10/MainTest.java:35]: these still need to be changed from "attributes" to "attribute" since we are only discussing one of the attributes

**Review by SamamthaH — CHANGES_REQUESTED** [review]

File: src/oop/special_class_method/hard/q10/QuestionText.md [hunk 25]
```diff
@@ -0,0 +1,37 @@
...
+
+- Attributes:
+    - A `String` attribute named **awardRecipient** which is not accessible from other classes
+    - A `String` attribute named **awardCategory** which is not accessible from other classes
+
+- Constructors:
+    - A constructor which is accessible from other classes and accepts two `String` parameters to initialize the
+      **awardRecipient** and **awardCategory** attributes in that order.
+
+- Methods:
```
Comment (Hedgemon4) [inline src/oop/special_class_method/hard/q10/QuestionText.md:15]: Please indicate the return type of the set methods

**Review by Hedgemon4 — CHANGES_REQUESTED** [review]

**Review by Hedgemon4 — APPROVED** [review]
